### PR TITLE
Gfycat Support

### LINKED ISSUE #23: Support Gfycat
**Is your feature request related to a problem? Please describe.**
N/A

**Describe the solution you'd like**
[Gifcat](https://gfycat.com/) is one of the biggest Gif sites, and has an API available. We should support Gyfcat as a provider backend.

The solution MUST follow the patterns used elsewhere in the project. In particular, the [Strategy Pattern](https://refactoring.guru/design-patterns/strategy) ([TS example](https://refactoring.guru/design-patterns/strategy/typescript/example)) MUST be used as in the GiphyStrategy. Ideally, the Client would be implemented as a [Singleton](https://refactoring.guru/design-patterns/singleton) ([TS example](https://refactoring.guru/design-patterns/singleton/typescript/example)) too.

**Describe alternatives you've considered**
N/A

**Additional context**
From a cursory glance, their API looked more complex than Giphy and Tenor's.
An suitable NPM package may already be available for Gfycat since they are so popular.

### PR DESCRIPTION
**Pull request by sethetter**
Resolves #23 

The tests don't seem to quite be working yet, and there's a bit more to consider for the auth flow, right now it's just accepting a raw auth token. We could do the oAuth flow in the constructor if we change the expected signature of the constructor input.

Thoughts?

### CODE REVIEW

**Comment by sethetter** [comment]
Tests passing now!

**Comment by sethetter** [comment]
Ah I see I missed the `GifStrategy` implementation, I'll get that now.

**Review by jamescallumyoung — APPROVED** [review]

**Comment by sethetter** [comment]
@jamescallumyoung I was playing with accepting a client ID and secret for the client constructor, and requesting an auth token, but then the constructor would need to be async. Otherwise the auth call could be made before a search or trending call, only if `authToken` is missing.

I wasn't sure how consistent you wanted to keep the client constructor APIs though. Any thoughts?

**Comment by jamescallumyoung** [comment]
Hi @sethetter, thanks for the contribution! Wow, this is spot on! I didn't expect to come back to such a complete solution the very next day! 😆👍

I approved the changes but I can see you made some changes a few minutes ago. I'll hold off on reviewing until you're done but it looks great so far 👍 

> I was playing with accepting a client ID and secret for the client constructor [...]. I wasn't sure how consistent you wanted to keep the client constructor APIs though. Any thoughts?

Idealy they would function _fairly_ similarly so the consumer can understand them all easily. That said, I'm happy for the Clients to vary quite a lot, as neccessary. The Strategy pattern in use means that the Clients can all be implemented any way necessary, since the GifProvider just consumes the Strategies. (There's links to a description of the pattern in #23 if you're not familiar with it.)

If possible, it would be nice to support *both* the Bearer Token and clientID+secret formats to give the consumer some flexibility.

**Comment by jamescallumyoung** [comment]
@sethetter, I converted this PR to a Draft. When it's ready for review, you can just click the Ready for Review button near the bottom of the page 👍

**Comment by sethetter** [comment]
@jamescallumyoung Actually it looks like you can use both the search and trending API endpoints _without_ any sort of auth!

I'm going to make the bearer token optional, just in case someone wants to build on that in the future. But I tested it with the script below and it seems to work fine!

```
const { GifProviderBuilder, GfycatStrategy, GfycatClient } = require('./dist');

async function main() {
  const builder = new GifProviderBuilder();
  const strat = new GfycatStrategy(GfycatClient.init().some());

  builder.addStrategy(strat);

  const provider = builder.build();

  const gifs = await provider.search("superman");

  console.log(gifs);
}

main().then(() => console.log('done!'));
```

**Comment by jamescallumyoung** [comment]
I just looked at the API docs and it does seem you can make search and trending requests without Auth. It makes sense to have an optional API Key then. I'll test this now 👍

Also, there's some rebase/merge issues (I suspect due to the merge commit). I'll fix those locally too since I also want to update the commit messages to Symantic Commit Messages. (I started to try this out lately so it's not in the CONTRIBUTING yet.)

**Comment by sethetter** [comment]
@jamescallumyoung could also squash and merge the PR with one nice semantic commit? Just a thought!

**Comment by jamescallumyoung** [comment]
@sethetter, could you please [allow maintainers to push to your fork PR branch](https://docs.github.com/en/free-pro-team@latest/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork). This has to be enabled in your PR (in your fork repo) for me to push the changes there.

**Comment by sethetter** [comment]
It's enabled already!

**Comment by jamescallumyoung** [comment]
@sethetter Hmm, I can't push to your branch. Could you double check?

**Comment by sethetter** [comment]
Screen Shot 2020-10-24 at 10 51 55 AM

It's definitely checked 🤔

**Comment by sethetter** [comment]
@jamescallumyoung are you pushing to the `23-gfycat-support` branch?

**Comment by jamescallumyoung** [comment]
@sethetter Okay, might be an issue on my side. I'll check it out...

**Comment by sethetter** [comment]
@jamescallumyoung I'll add you as a contributor on my fork, that should cover it.

File: src/clients/gfycat/utils.ts
Comment (jamescallumyoung) [inline src/clients/gfycat/utils.ts:19]: I changed these lines since we were previously using the non-existant `poster_url` (it's actually posterUrl).

I removed the posterUrl completely though as that returns a static image. Since the other Strategies return low-quality Gifs in the thumbnailUrl field, I did the same here.

The long `||` chain is required to handle the fact that Gfycat can return objects with sometimes non-existant fields.

Comment (jamescallumyoung) [reply src/clients/gfycat/utils.ts:19]: This should have been caught in the tests. I'll update them seperately to keep this commit clean.

**Review by jamescallumyoung — APPROVED** [review]
LGTM 👍

I rebased and squashed to a single semantic commit.
I also added a fix for the thumbnailUrl not being set (I'll add a comment to the changed lines).

Thanks for the contribution @sethetter! Collaborating on OSS can sometimes be a pain but this was a breeze; great work, it's very much appreciated!

File: src/clients/gfycat/utils.ts
Comment (sethetter) [reply src/clients/gfycat/utils.ts:19]: Thanks for the quick feedback on all this! Glad the process was nice and smooth 👍

**Pull request merged**
